pycozmo/activity.py

```diff
@@ -14,30 +16,389 @@
+            strategy=data['activityStrategy']['type'])
+
+
+class SocializeActivity(Activity):
+    def __init__(self,
+                 behavior_chooser: BehaviorChooser,
+                 interlude_chooser: BehaviorChooser,
+                 max_face_iterations: int,
+                 required_objectives: List[Objective],
+                 *args, **kwargs):
+        self.behavior_chooser = behavior_chooser
+        self.interlude_chooser = interlude_chooser
+        self.max_face_iterations = max_face_iterations
+        self.required_objectives = required_objectives
+
+        super().__init__(*args, **kwargs)
+
+    def choose(self):
+        if self.behavior_chooser:
+            return self.behavior_chooser.choose()
+
+    @classmethod
+    def from_json(cls, data: Dict):
+        return cls(
+            behavior_chooser=BehaviorChooser.from_json(data['behaviorChooser']),
+            interlude_chooser=BehaviorChooser.from_json(data['interludeBehaviorChooser']),
+            max_face_iterations=data['maxNumFindFacesSearchIterations'],
+            required_objectives=[Objective.from_json(d) for d in data['requiredObjectives']],
+            activity_id=data['activityID'],
+            activity_type=data['activityType'],
+            strategy=data['activityStrategy']['type'])
+
+
+class NeedsActivity(Activity):
+    def __init__(self,
+                 behavior_chooser: BehaviorChooser,
+                 *args, **kwargs):
+        self.behavior_chooser = behavior_chooser
+        super().__init__(*args, **kwargs)
+
+    def choose(self):
+        if self.behavior_chooser:
+            return self.behavior_chooser.choose()
+
+    @classmethod
+    def from_json(cls, data: Dict):
+        return cls(
+            behavior_chooser=BehaviorChooser.from_json(data['behaviorChooser']),
+            activity_id=data['activityID'],
+            activity_type=data['activityType'],
+            strategy=data['activityStrategy']['type'])
+
+
+def from_dict(info: Dict) -> Activity:
+    if info['activityType'] == 'VoiceCommand':
+        return VoiceCommandActivity.from_json(info)
+
+    elif info['activityType'] == 'BehaviorsOnly':
+        return BehaviorsActivity.from_json(info)
+
+    elif info['activityType'] == 'Feeding':
+        return FeedingActivity.from_json(info)
+
+    elif info['activityType'] == 'Freeplay':
+        return FreeplayActivity.from_json(info)
+
+    elif info['activityType'] == 'Sparked':
+        return SparkedActivity.from_json(info)
+
+    elif info['activityType'] == 'BuildPyramid':
+        return PyramidActivity.from_json(info)
+
+    elif info['activityType'] == 'Socialize':
+        return SocializeActivity.from_json(info)
+
+    elif info['activityType'] == 'NeedsExpression':
+        return NeedsActivity.from_json(info)
+
+    else:
+        return Activity(
+            activity_id=info['activityID'],
+            activity_type=info['activityType'],
+            strategy=info['activityStrategy']['type']
+        )
```

Comment:
👍

Reply:
There is one thing I wanted to ask. In this class, you added some slots. Should I include slots for the non-optional variables in the different activities?

Reply:
`__slots__` is a way to restrict what fields can be used in a class. Feel free to remove them for now for activities.